### DISCUSSION TITLE
[Si4432] - Receiving door sensor data in FSK mode

### DISCUSSION THREAD

**Discussion by Arthurk12** [jgromes/RadioLib#442] [Q&A] [answered]
Hello guys,

Recently I bought some comercial 433.92 MHz door sensors from a traditional brand here in Brazil, its model is **Intelbras XAS 4010 Smart**. This sensor has 2 operating modes, OOK and FSK and uses a _SI4010_. I already manage to receive the door sensor data on OOK mode(using another RF module) but now I want to go further and receive the sensor's data on FSK mode with my _SI4432_ module + ESP8266.

![Sinal1](https://user-images.githubusercontent.com/32987232/149688870-f7952d8e-a9e8-4a77-bab6-f9eb51b78229.png)

Using a SDR I was able to discover the following parameters:
1. Bit rate is 1.6kbps (despite the user manual saying it is 2kBps)
2. The frequency deviation is about 30 KHz (i'm not shure about the receiver bandwidth, if someone could explain me)

Analyzing the captured transmissions with URH I discovered:
![image](https://user-images.githubusercontent.com/32987232/149688541-32e27fb8-6926-4528-9d86-89cacea62e39.png)
![image](https://user-images.githubusercontent.com/32987232/149686162-d2ccbd85-e7e6-4f7a-9df7-db5b87e0a999.png)

- The first 24 bits are the preamble
- The following 16 bits are the sync word, which are `0x2D, 0xD4`
- The last 24 bits are the data
- The sensor repeat each transmission 8 times.

The problem is that when I run the [receive_interrupt](https://github.com/jgromes/RadioLib/blob/master/examples/Si443x/Si443x_Receive_Interrupt/Si443x_Receive_Interrupt.ino) example for the _SI4432_ with the above parameters, I'm receiving different/random data on each transmission. 
Below is the serial monitor output of repeted transmissions:
```
[Si4432] Initializing ... success!
Success setting syncWord!
[Si4432] Starting to listen ... success!
[Si4432] Received packet!
[Si4432] Data:			6E6E6E6E6E6E6E6E
[Si4432] Received packet!
[Si4432] Data:			1313131313131313
[Si4432] Received packet!
[Si4432] Data:			9393939393939393
[Si4432] Received packet!
[Si4432] Data:			D2D2D2D2D2D2D2D2
[Si4432] Received packet!
[Si4432] Data:			5CE518D5555516E
[Si4432] Received packet!
[Si4432] Data:			5CE3AC1D5555516E
[Si4432] Received packet!
[Si4432] Data:			5CE558D5555516E
[Si4432] Received packet!
[Si4432] Data:			5CE3ACD5555516E
[Si4432] Received packet!
[Si4432] Data:			5CE58E8E8E8E8E8E
[Si4432] Received packet!
[Si4432] Data:			8181818181818181
[Si4432] Received packet!
[Si4432] Data:			5CE5333333
[Si4432] Received packet!
[Si4432] Data:			9090909090909090
[Si4432] Received packet!
[Si4432] Data:			8B8B8B8B8B8B8B8B
[Si4432] Received packet!
[Si4432] Data:			A4A4A4A4A4A4A4A4
[Si4432] Received packet!
[Si4432] Data:			4949494949494949

```

Any help is appreciated!

**Comment by jgromes** [comment]
The receiver bandwidth just specifies width of a filter on the receiver input - it should be wide enough for your signal to fit, so if you're receiving data, then that's fine.

By default, the module expects this structure: preamble-sync-length-payload.
However, your module does not transmit the length byte, because all transmissions have fixed length. Si4432 does support fixed packet length transmissions, however, it seems I was lazy and didn't implement it. I'll try to add it in the next couple of days.

**Reply by jgromes** [reply, marked as answer]
@Arthurk12 fixed packet length mode was added, you can enable it by calling `fixedPacketLengthMode(len);`. Let me know if that helped.

**Reply by Arthurk12** [reply]
Thank you for the help, @jgromes !

I've tested it and kind of works. I'm saying it because when the fixed packet length is set to 3 bytes(which are the length of the data sent by the sensor) the received bytes are the first one repeated 3 times. For example, one of the sensors' data is `da5ce5`, but the received bytes are `dadada`.

So what I did and worked(although not consistently) is to set the fixed packet length to 50 bytes and calling the `readData` only for 3 bytes. This sometimes work and sometimes not, I couldn't notice any pattern. Below is the serial monitor output:
```
17:17:03.148 -> [Si4432] Received packet!
17:17:03.188 -> [Si4432] Data: DADADA
17:17:09.167 -> [Si4432] Received packet!
17:17:09.208 -> [Si4432] Data: DADADA
17:17:15.198 -> [Si4432] Received packet!
17:17:15.198 -> [Si4432] Data: DADADA
17:17:19.178 -> [Si4432] Received packet!
17:17:19.217 -> [Si4432] Data: DA5CE3
17:17:24.167 -> [Si4432] Received packet!
17:17:24.212 -> [Si4432] Data: DA5CE5
17:17:25.168 -> [Si4432] Received packet!
17:17:25.207 -> [Si4432] Data: DADADA
17:17:29.174 -> [Si4432] Received packet!
17:17:29.217 -> [Si4432] Data: DADADA
17:17:34.168 -> [Si4432] Received packet!
17:17:34.207 -> [Si4432] Data: DA5CE5
17:17:39.188 -> [Si4432] Received packet!
17:17:39.188 -> [Si4432] Data: DADADA
17:17:43.198 -> [Si4432] Received packet!
17:17:43.198 -> [Si4432] Data: DADADA
17:17:47.178 -> [Si4432] Received packet!
17:17:47.217 -> [Si4432] Data: DADADA
17:17:51.207 -> [Si4432] Received packet!
17:17:51.207 -> [Si4432] Data: DA5CE5
17:17:56.188 -> [Si4432] Received packet!
17:17:56.228 -> [Si4432] Data: DADADA
17:17:58.197 -> [Si4432] Received packet!
17:17:58.197 -> [Si4432] Data: DADADA
17:18:00.172 -> [Si4432] Received packet!
17:18:00.207 -> [Si4432] Data: DA5CE5
```

Another thing I noticed and thought could be the missing piece is that the `getPacketLength` function was not touched and so it keeps reading the packet length from `RADIOLIB_SI443X_REG_RECEIVED_PACKET_LENGTH` register when it should be read from `RADIOLIB_SI443X_REG_TRANSMIT_PACKET_LENGTH` according to the datasheet. I ventured doing this change, based on how it is done on other [modules](https://github.com/jgromes/RadioLib/blob/19109bc7d84e462b1309b7e05332a8f112b4dcb4/src/modules/RF69/RF69.cpp#L697-L708) but had no success(still receiving the first data byte repeated). Anyway the function ended up like this: 
```
size_t Si443x::getPacketLength(bool update) {
  if(!_packetLengthQueried && update) {
    if (_packetLengthConfig == RADIOLIB_SI443X_FIXED_PACKET_LENGTH_ON) {
      _packetLength = _mod->SPIreadRegister(RADIOLIB_SI443X_REG_TRANSMIT_PACKET_LENGTH);
    } else {
      _packetLength = _mod->SPIreadRegister(RADIOLIB_SI443X_REG_RECEIVED_PACKET_LENGTH);
    }
    _packetLengthQueried = true;
  }

  return(_packetLength);
}
```

Do you have any idea what could be wrong?

**Reply by Arthurk12** [reply]
For some magic reason, when I put `delay(200);` [below this line](https://github.com/jgromes/RadioLib/blob/19109bc7d84e462b1309b7e05332a8f112b4dcb4/examples/Si443x/Si443x_Receive_Interrupt/Si443x_Receive_Interrupt.ino#L94) consistent values being received.

ps: still using 50 bytes of fixed packet length and reading just the first 3 bytes.

**Reply by jgromes** [reply]
I would try using packet length of 3 instead of 50. The Rx interrupt is generated only after a complete packet is received.

And you're right, I forgot to update the `getPacketLength` method. Feel free to submit a PR with your modification, it looks OK and seems to be working.
